web/pandas/pdeps/0006-ban-upcasting.md

```diff
@@ -51,17 +51,24 @@ In [11]: ser[2] = '2000-01-04x'  # almost certainly a typo - but pandas doesn't
 ```
 
 The scope of this PDEP is limited to setitem-like operations which would operate inplace, such as:
-- ``ser[0] == 2``;
-- ``ser.fillna(0, inplace=True)``;
-- ``ser.where(ser.isna(), 0, inplace=True)``
+- ``ser[0] = 2.5``;
+- ``ser.fillna('foo', inplace=True)``;
+- ``ser.where(ser.isna(), 'foo', inplace=True)``
+- ``ser.iloc[0] = 2.5``
+- ``ser.loc[0] = 2.5``
+- ``ser[:] = 2.5``
 
 There may be more. What is explicitly excluded from this PDEP is any operation would have no change
 of operating inplace to begin with, such as:
 - ``ser.diff()``;
 - ``pd.concat([ser, other])``;
-- ``ser.mean()``.
+- ``ser.mean()``;
+- ``df.loc[0, 'col1'] = 2.5`` (if ``df`` is not a single block).
 
-These would keep being allowed to change Series' dtypes.
+These would keep being allowed to change Series' dtypes. Note that setting element of a column of a
+``DataFrame`` would not raise, as that sets the elements in a new block manager (rather than in the
+original one),
+see https://github.com/pandas-dev/pandas/blob/4e4be0bfa8f74b9d453aa4163d95660c04ffea0c/pandas/core/internals/managers.py#L1361-L1362.
```

Comment:
This means the user has to be aware of the block manager?  I think this will be really confusing

Reply:
They wouldn't, no - I just intended this as an explanation to developers

Users just need to know that, for Series setitem-like operations, incompatible dtypes will raise

Reply:
I think this differentiation of doing a setitem-like operation on  a `Series` versus changing a column of a `DataFrame`, which is a `Series`, can be really confusing.  Here's current behavior:
```python
>>> s=pd.Series([1,2,3])
>>> df=pd.DataFrame({"x": s})
>>> df.dtypes
x    int64
dtype: object
>>> s2=df["x"]
>>> s2.iloc[1] = 2.5
>>> df
     x
0  1.0
1  2.5
2  3.0
```

So here, when modifying `s2`, in  your proposal, you would raise an exception, but if I did `df.loc[1, "x"] = 2.5`, there would be no exception, but the underlying change to `df` would be the same ?

You're asking users to know where the `Series` came from.  If it is a reference to a column of a `DataFrame`, then the behavior is different than if the `Series` was created separately.

I'm fine with the behavior you are proposing, but I think that it needs to happen for `DataFrame` as well.

Reply:
that's a good point - I've updated, thanks!